src/fifo.c

```diff
@@ -0,0 +1,338 @@
+ *  We need to keep track of the first & last slot used.  Contextually, we will only need
+ *  a single index - either the first slot used or the last slot used.  Based on context,
+ *  we can determine what is needed.
+ *
+ * Blocks are linked together in a chain.  If the list is empty, there are no blocks.
+ *  For non-empty lists, we will either have a single block OR a chain of blocks.
+ *
+ * For a SINGLE BLOCK containing (for example) 4 items, the layout looks like this:
+ *                 +--------+--------+--------+--------+--------+--------+--------+--------+
+ *  SINGLE BLOCK:  | slot 0 | slot 1 | slot 2 | slot 3 | slot 4 | slot 5 | slot 6 | next/  | 
+ *                 |  item  |  item  |  item  |  item  |   -    |   -    |   -    | lastIdx|
+ *                 +--------+--------+--------+--------+--------+--------+--------+--------+
+ *                                                ^
+ *                                             lastIdx (3)
+ *  In single blocks, the items are always shifted so that the first item is in slot 0.
+ *  We need to keep track of the lastIdx so that we will know where to push the next item.
+ *  The last index is stored in the final 3 bits of the (unused) next pointer
+ *
+ * When MULTIPLE BLOCKS are chained together, items will be popped from the first block, and
+ *  pushed onto the last block.  All blocks in the middle are full(*).  In the first block, we keep
+ *  the firstIdx (so we know where to pop) ... on the last block, we keep lastIdx (so we know
+ *  where to push).
+ * 
+ * (*) While blocks in the middle of a chain are generally full, the Fifo supports O(1) joining of
+ *     two lists.  In this case, a block at the join point may not be full.  In this case, it will
+ *     look like the FIRST BLOCK below, with the first index stored in the indexing bits.
+ *
+ * Example FIRST BLOCK with 2 items remaining:
+ *                 +--------+--------+--------+--------+--------+--------+--------+--------+
+ *  FIRST BLOCK:   | slot 0 | slot 1 | slot 2 | slot 3 | slot 4 | slot 5 | slot 6 | next/  | 
+ *                 |   -    |   -    |   -    |   -    |   -    |  item  |  item  |firstIdx|
+ *                 +--------+--------+--------+--------+--------+--------+--------+--------+
+ *                                                                  ^
+ *                                                              firstIdx (5)
+ * Example LAST BLOCK with 3 items pushed so far:
+ *                 +--------+--------+--------+--------+--------+--------+--------+--------+
+ *  LAST BLOCK:    | slot 0 | slot 1 | slot 2 | slot 3 | slot 4 | slot 5 | slot 6 | next/  | 
+ *                 |  item  |  item  |  item  |   -    |   -    |   -    |   -    | lastIdx|
+ *                 +--------+--------+--------+--------+--------+--------+--------+--------+
+ *                                        ^
+ *                                    lastIdx (2)
+ */
+typedef struct FifoBlock {
```

Comment:
how about 32-bit system, I think we should align FifoBlock to 8 byte explicitly to guarantee 3 unused bits in `next`
```suggestion
typedef struct __attribute__((aligned(8))) FifoBlock {
```

Reply:
The `aligned` attribute is applicable to compiler configured memory.  Specifically globals and stack.  It can also affect the padding within a structure.  But it doesn't affect the alignment for a block of heap memory (returned by `malloc`).

`malloc`, by definition, is required to return blocks which are aligned for the largest type that can fit in the block.  `malloc` doesn't know what you will be storing in the block.  So, a `malloc(5)` is guaranteed to be (at least) 4-byte aligned (since is might hold an integer at offset 0).  Any block 8-bytes or larger is guaranteed to be (at least) 8-byte aligned.  On architectures where there are datatypes requiring 16 byte alignment, any malloc operation for 16 or more bytes would have to be 16 byte aligned.

Even on a 32-bit system, there are types requiring 8 byte alignment.  We can be assured that malloc will always return a block that's (at least) 8 byte aligned.

Reply:
Worth mentioning: For allocators which lack a `malloc_size()`, zmalloc didn't respect the alignment requirement, since it used to add a prefix of `sizeof(size_t)` bytes before the returned address (`zmalloc()` returns `ptr+PREFIX_SIZE`) which on 32bit platforms i 4. This is now changed so PREFIX_SIZE is at least 8, which guarantees the returned pointer is 8 bytes aligned.

This was not always the case, but now it is, to allow a similar pointer bit trick in dict.c.